crates/services/sync/src/service.rs

```diff
@@ -1,61 +1,232 @@
+    params: Params,
+    ports: ports::Ports<P, E, C>,
 }
 
-pub struct Service {
-    join: Mutex<Option<JoinHandle<()>>>,
-    sender: mpsc::Sender<SyncMpsc>,
+/// The task that waits for the service to continue running.
+pub struct Task {
+    sync_task: Option<JoinHandle<()>>,
+    import_task: Option<JoinHandle<anyhow::Result<bool>>>,
+    kill_switch: KillSwitch,
 }
 
-impl Service {
-    pub async fn new(_config: &Config) -> anyhow::Result<Self> {
-        let (sender, _receiver) = mpsc::channel(100);
-        Ok(Self {
-            sender,
-            join: Mutex::new(None),
-        })
+impl<P, E, C> TaskSetup<P, E, C>
+where
+    P: ports::PeerToPeerPort + Send + Sync + 'static,
+    E: ports::ExecutorPort + Send + Sync + 'static,
+    C: ports::ConsensusPort + Send + Sync + 'static,
+{
+    /// Create a new task setup.
+    pub fn new(
+        height_stream: BoxStream<'static, BlockHeight>,
+        state: State,
+        params: Params,
+        ports: ports::Ports<P, E, C>,
+    ) -> Self {
+        let state = SharedMutex::new(state);
+        Self {
+            height_stream,
+            state,
+            params,
+            ports,
+        }
     }
+}
+
+#[async_trait::async_trait]
+impl RunnableTask for Task {
+    async fn run(
+        &mut self,
+        watcher: &mut fuel_core_services::StateWatcher,
+    ) -> anyhow::Result<bool> {
+        use futures::future::{
+            select,
+            Either,
+        };
 
-    pub async fn start(
-        &self,
-        _p2p_block: mpsc::Receiver<BlockGossipData>,
-        // TODO: re-introduce this when sync actually depends on the coordinator
-        // _bft: mpsc::Sender<BftMpsc>,
-        _block_importer: mpsc::Sender<()>,
-    ) {
-        let mut join = self.join.lock();
-        if join.is_none() {
-            *join = Some(tokio::spawn(async {}));
+        // Check we still have tasks to run otherwise return false.
+        match (&mut self.sync_task, &mut self.import_task) {
+            (Some(sync_task), Some(import_task)) => {
+                // Select over the task runner state, the sync task and the import task.
+                // Match on either the sync task or the import task completing.
+                // If the task runner state changes then we check it below.
+                if let Either::Right((join_handles, _)) =
+                    select(watcher.changed().boxed(), select(sync_task, import_task))
+                        .await
+                {
+                    match join_handles {
+                        // Sync task has completed.
+                        Either::Left((sync_result, import_task)) => match sync_result {
+                            // Sync task has panicked.
+                            Err(e) if e.is_panic() => {
+                                // Kill the import task and resume the panic.
+                                self.kill_switch.kill_all();
+                                let _ = import_task.await;
```

Comment:
The main reason for handling these results I'd assume is that the task is being killed anyway? Is there any chance for a messy shutdown with errors being reason for panic here?

Reply:
Yeh I thought about that but you can't really combine two panics (well not that I know of) so we don't really have a choice